File: 931/solution.go
package _931

func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func minFallingPathSum(matrix [][]int) int {
	n := len(matrix)
	dp := make([][]int, n)
	for i := range dp {
		dp[i] = make([]int, n)
	}
	copy(dp[0], matrix[0])
	for i := 1; i < n; i++ {
		for j := 0; j < n; j++ {
			tmp := dp[i-1][j]
			if j > 0 {
				tmp = min(tmp, dp[i-1][j-1])
			}
			if j < n-1 {
				tmp = min(tmp, dp[i-1][j+1])
			}
			dp[i][j] = matrix[i][j] + tmp
		}
	}
	res := dp[n-1][0]
	for i := 1; i < n; i++ {
		res = min(res, dp[n-1][i])
	}
	return res
}

File: 931/solution.py
import sys
from typing import List


class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        dp = [sys.maxsize] * (n + 2)
        for i in range(n):
            dp[i + 1] = matrix[0][i]
        for i in range(1, n):
            pre = dp[1]
            for j in range(n):
                tmp = dp[j + 1]
                dp[j + 1] = matrix[i][j] + min(pre, dp[j + 1], dp[j + 2])
                pre = tmp
        return min(dp)
